### ISSUE TITLE
explicit CORS origin

### ISSUE THREAD

**Issue by FND** [neo4j/neo4j#5628]
Neo4j [currently](https://github.com/neo4j/neo4j/blob/80149338b7fbe3f6df7df6ec95312af742320195/community/server/src/main/java/org/neo4j/server/rest/web/AllowAjaxFilter.java#L43) responds with `Access-Control-Allow-Origin: *` to allow CORS requests.

However, browsers' [fetch](https://fetch.spec.whatwg.org) implementations (see below) require an explicit origin when passing credentials:

> Fetch API cannot load http://localhost:7474/. A wildcard '*' cannot be used in the 'Access-Control-Allow-Origin' header when the credentials flag is true. Origin 'http://localhost:8000' is therefore not allowed access.

(Chrome)

> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost:7474/. (Reason: CORS header 'Access-Control-Allow-Origin' does not match '*').

(Firefox)

The problem goes away when the server response mirrors the request origin (i.e. `Access-Control-Allow-Origin: $REQUEST_HEADERS["Origin"]` - though that's not actually more secure, of course).

---

`fetch` is essentially a modern reinvention of XHR. In my case, I tried to POST a Cypher query from an application at localhost:8000 to Neo4j at localhost:7474:

``` javascript
fetch("http://localhost:7474/db/data/transaction/commit", {
     credentials: "include",
     method: "POST",
     headers: {
         "Content-Type": "application/json",
         Accept: "application/json"
     },
     body: JSON.stringify(payload)
 }).then(res => res.json()).then(console.log);
```

**Comment by ikwattro** [comment]
```
This is an experimental technology
Because this technology's specification has not stabilized, check the compatibility table for the proper prefixes to use in various browsers. Also note that the syntax and behavior of an experimental technology is subject to change in future versions of browsers as the spec changes.
```

https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

IMO, this doesn't make sense to make any changes to support unstable APIs.

**Comment by sideshowbarker** [comment]
> > Neo4j currently responds with Access-Control-Allow-Origin: \* to allow CORS requests.
> > 
> > However, browsers' fetch implementations (see below) require an explicit origin when passing credentials
> 
> IMO, this doesn't make sense to make any changes to support unstable APIs.

That part of the spec is not unstable. CORS has been around since 2007—though by a different name earlier (see http://www.w3.org/TR/2007/WD-access-control-20070215/) and as _CORS_ since 2009 http://www.w3.org/TR/2009/WD-cors-20090317/

So the CORS requirements are at least 6-8 years old at this point, with server-side support deployed all over the place (e.g., in libraries for all kinds of server-side runtimes) and the CORS restrictions are essentially locked in as-is to browsers now.

I guess the part that’s a bit confusing is that the current spec where the CORS requirements live is the Fetch spec, which is a mix of sorta documenting stable technologies like CORS, along with defining the relatively newer actual `fetch(…)` API that replaces XHR.

In summary, I believe the OP is correct here in pointing out that for the credentials-included case, neo4j should be responding with the explicit requesting origin, not with the the `*` wildcard.

**Comment by annevk** [comment]
And also with the `Access-Control-Allow-Credentials: true` header. Though note that this means you give the other origin access to all credentialed data.

**Comment by ikwattro** [comment]
I didn't said CORS is not stable, only `fetch` which is not yet compatible in all browsers, especially for streaming https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API#Browser_compatibility

On the other hand, CORS is working with a neo4j server, I can make CORS requests (without fetch) to a local neo4j instance from a web application without problems.

**Comment by FND** [comment]
I agree that the core issue is with the fetch API being more restrictive than XHR when credentials are included. However, it seems to me that using an explicit origin is a good idea anyway, regardless of fetch:

> The string "*" cannot be used for a resource that supports credentials.

http://www.w3.org/TR/cors/#resource-requests

FWIW, in my case, during local development, my Neo4j instance doesn't actually require credentials. However, since it seems like a bad idea to introduce a special case to omit `{ credentials: "include" }` in this scenario, fetch still expects that header value to not be "*".

**Comment by annevk** [comment]
@FND that sounds inaccurate. If you use `credentials: "same-origin"` it should behave the same as XMLHttpRequest. XMLHttpRequest with `withCredentials = true` should behave the same as `credentials: "include"`.

**Comment by FND** [comment]
Thanks @annevk - in that case, I assume that @ikwattro would run into a similar error with non-fetch CORS requests as soon as credentials are used (via XHR's `withCredentials = true`).

**Comment by ikwattro** [comment]
I don't have any specific xhr settings except Authorization header.

**Comment by wanderview** [comment]
> I agree that the core issue is with the fetch API being more restrictive than XHR when credentials are included.

I'm not sure this is an accurate statement.  For example, Firefox also blocks XHR when `Access-Control-Allow-Origin: *` with `Access-Control-Allow-Credentials: true`.  See this test:

https://dxr.mozilla.org/mozilla-central/source/dom/security/test/cors/test_CrossSiteXHR.html#758

**Comment by cleishm** [comment]
On my todo list for a while has been to sit down and (re)read the CORS specs to understand the implications on Authorization headers. If anyone feels like doing that research and writing it up, that would certainly help get this moved forward.

The security concerns are to ensure that credentials are not exposed outside of the requesting page in the browser - e.g. assuming a user has Neo4j Browser open, another web page in a different browser window/tab should not be able to access the credentials and make requests to Neo4j Server.

**Comment by sideshowbarker** [comment]
> On my todo list for a while has been to sit down and (re)read the CORS specs to understand the implications on Authorization headers.

If you do that, please make sure the spec you read is https://fetch.spec.whatwg.org/ I think that’s actually the _only_ one you need to read for CORS requirements. All other specs for CORS published elsewhere are either obsoleted/superseded by it or are some subset of it. (@annevk and correct me if I’m missing something.)

**Comment by technige** [comment]
Given the age of this thread, the lack of recent activity and the fact that we'd now recommend Bolt connections over HTTP, I'm going to close down this issue. If any contributors feel this is still something worth considering, please feel free to follow up.
